### PR TITLE
Add Collect wrapper

### PR DESCRIPTION
**Pull request by onelxj**

### CODE REVIEW

File: test/collect.cc [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  auto s = [&]() {
+    return Iterate(v)
+        | Collect<std::set<int>>([](int x) {
+             return x != 5;
+           });
+  };
+
+  std::set<int> res = *s();
+
+  EXPECT_EQ(12, *res.begin());
```
Comment (FolMing) [inline test/collect.cc:43]: `res.at(0)`

File: test/collect.cc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  auto s = [&]() {
+    return Iterate(v)
+        | Collect<std::vector<int>>([](int x) {
+             return x != 5;
+           });
+  };
+
+  std::vector<int> res = *s();
+
+  EXPECT_EQ(12, res[0]);
```
Comment (FolMing) [inline test/collect.cc:27]: `res.at(0)` to also check if you're not OOB.

**Review by FolMing — CHANGES_REQUESTED** [review]

File: stout/collect.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        stream.Next();
+      })
+      .ended([](auto& data, auto& k) {
+        k.Start(std::move(data.content_));
+      });
+}
+
+template <typename Container>
+auto TakeFirstAsContainer() {
+}
```
Comment (benh) [inline stout/collect.h:69]: Dead code?

File: stout/collect.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename Container, typename F>
+auto Collect(F func) {
+  using T = typename Container::value_type;
+
+  struct DataContainer {
+    std::function<bool(T)> func_;
+    Container content_;
```
Comment (benh) [inline stout/collect.h:47]: Underscores only for private members of a class/struct.

**Review by benh — COMMENTED** [review]
This PR is implementing both 'Collect' and 'Filter'. Let's factor these into two separate things. Ultimately someone should be able to do:

```c++
auto e = []() {
  return Iterate(std::vector<int>{1, 2, 3, 4, 5})
      | Filter([](int& i) {
        return i % 2;
      })
      | Collect<std::vector<int>>();
};

std::vector<int> evens = *e();
```

Make sense? Let's just implement 'Collect' in this PR and then we can do 'Filter' in another PR.

**Comment by FolMing** [comment]
This PR also adds libcurl rules. Pls remove

**Pull request closed without merging**